### ISSUE TITLE
[Question] Real-world online / continual learning from trained model

### ISSUE THREAD

**Issue by manjavacas** [DLR-RM/stable-baselines3#2100]
### ❓ Question

My objective is to solve the loss of performance of an agent after the properties of the environment on which it was initially trained have been modified. The following is an example I am currently implementing:

- Suppose I start by training an off-policy agent such as SAC in, for example, the `Pendulum-v1` Gymnasium environment:

```python
train_env = gym.make('Pendulum-v1')

base_model = SAC('MlpPolicy', train_env)
base_model.learn(total_timesteps=200_000)

base_model.save('base_model')
```
- Once trained, I create a 'production environment' where instead of using the default gravity value (`g=9.81`), I change it to `g=11`. My objective is to:

1. Collect experience over _N_ timesteps (e.g., until the replay buffer is full) in this new environment.
2. Train the agent with the collected experience (not from scratch, but from its current state).
3. Repeat iteratively until the agent has adapted to the new environment.

Something like this:

```python
production_env = gym.make('Pendulum-v1', g=11)

model = SAC.load('base_model')

for _ in range(10):
    
    # 1) gather experience until replay buffer is full
    # 2) model train update

mean_reward, std_reward = evaluate_policy(base_model, production_env, n_eval_episodes=10, deterministic=True)
print(f"PRODUCTION Environment -- Mean reward: {mean_reward} ± {std_reward}")

model.save('tuned_model')
```

My problem is that I don't know if StableBaselines3 allows to implement this kind of continuous learning through its API (i.e. accessing the buffer, training on the collected data...). I've noticed that it might be possible to use `collect_rollouts` along with `model.train()` for this process, but I'm not entirely sure how to implement it. Could someone provide an example of how to use these functions together?

I'm aware that I could declare the new environment and re-run `model.learn()`, but I am interested in being able to bring this into a non-simulated environment where data is obtained from real-world interaction.

Thanks in advance!

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [x] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [x] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
> but I am interested in being able to bring this into a non-simulated environment where data is obtained from real-world interaction.

Simulated or not simulated, gym is just an interface, see https://github.com/DLR-RM/stable-baselines3/issues/2082#issuecomment-2642550594 .
So the recommended way would be to define a Gym interface for real-world online training.

Otherwise, you need to re-implement the `learn()` loop manually (which involves re-implementing part of `collect_rollout()`):
https://github.com/DLR-RM/stable-baselines3/blob/656de97269c9e3051d3bbfb3f5f328d486867bd8/stable_baselines3/common/off_policy_algorithm.py#L327-L347

**Comment by araffin** [comment]
> but would it be correct to manually fill the buffer and subsequently call .train() after a certain number of steps? Something like this:

yes, but that's what `learn()` and `collect_rollout` is about normally (especially to have the correct normalized action, unnormalized obs in the buffer).

**Comment by manjavacas** [comment]
@araffin Thank you for the answer, so far it is clear to me 👍🏻
